// angular-web-worker/esm2015/angular/lib/client-web-worker.js
import { WorkerController } from 'angular-web-worker';
/**
 * Used to mock the behaviour of the native `Worker` class when a `WorkerClient` is set to run in the app and not in the worker script.
 * Controls the flow of messages to and from a `WorkerClient` and a `WorkerController`
 */
export class ClientWebWorker {
    /**
     * Creates a new instance of a `ClientWebWorker`
     * @param workerType the worker class
     * @param isTestClient whether the instance is used for testing which will then mock serialization
     */
    constructor(workerType, isTestClient) {
        this.isTestClient = isTestClient;
        this.messageBus = {
            onmessage: () => { },
            postMessage: (resp) => {
                this.onmessage(new MessageEvent('ClientWebWorker', { data: this.isTestClient ? this.serialize(resp) : resp }));
            }
        };
        this.controller = new WorkerController(workerType, this.messageBus);
    }
    /**
     * Returns instance of worker class
     */
    get workerInstance() {
        return this.controller.workerInstance;
    }
    /**
     * Message listener for a `WorkerClient`
     */
    onmessage(ev) {
    }
    /**
     * Sends messages triggered from a `WorkerClient` to a `WorkerController`
     */
    postMessage(resp) {
        this.messageBus.onmessage(new MessageEvent('ClientWebWorker', { data: this.isTestClient ? this.serialize(resp) : resp }));
    }
    /**
     * Unsubscribes from all subscriptions in the `WorkerController` and then destroys the controller
     */
    terminate() {
        this.controller.removeAllSubscriptions();
        this.controller = null;
    }
    /**
     * Used for testing to mock the serialization that occurs when native the postMessage or onmessage are used to communicate with a worker script
     * @param obj object to be serialised
     */
    serialize(obj) {
        return JSON.parse(JSON.stringify(obj));
    }
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    onerror(err) {
    }
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    addEventListener() {
    }
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    removeEventListener() {
    }
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    dispatchEvent(evt) {
        return true;
    }
}
