// angular-web-worker/esm2015/angular/lib/worker.module.js
import * as tslib_1 from "tslib";
var WorkerModule_1;
import { WorkerManager } from './worker-manager';
import { NgModule } from '@angular/core';
import { WorkerUtils, WorkerAnnotations } from 'angular-web-worker/common';
/**
 * Provides the `WorkerManager` service with the worker definitions passed into the static `forWorkers` method.
 * @example
 * imports: [
 *  WorkerModule.forWorkers([
 *    {worker: AppWorker, initFn: () => new Worker('./app.worker.ts', {type: 'module'})},
 *  ])
 * ]
 */
let WorkerModule = WorkerModule_1 = class WorkerModule {
    /**
     * Returns a module with a `WorkerManager` provider
     * @param workerDefinitions list of worker defintions which contain the worker class and an `initFn` function which is necessary for the
     * webpack `worker-plugin` to bundle the worker seperately.
     * @example
     * imports: [
     *  WorkerModule.forWorkers([
     *    {worker: AppWorker, initFn: () => new Worker('./app.worker.ts', {type: 'module'})},
     *  ])
     * ]
    */
    static forWorkers(workerDefinitions) {
        workerDefinitions.forEach((definition) => {
            if (!WorkerUtils.getAnnotation(definition.worker, WorkerAnnotations.IsWorker)) {
                throw new Error('WorkerModule: one or more of the provided workers has not been decorated with the @AngularWebWorker decorator');
            }
        });
        return {
            ngModule: WorkerModule_1,
            providers: [
                { provide: WorkerManager, useValue: new WorkerManager(workerDefinitions) }
            ]
        };
    }
};
WorkerModule = WorkerModule_1 = tslib_1.__decorate([
    NgModule()
], WorkerModule);
export { WorkerModule };
