// angular-web-worker/esm5/angular/lib/client-web-worker.js
import { WorkerController } from 'angular-web-worker';
/**
 * Used to mock the behaviour of the native `Worker` class when a `WorkerClient` is set to run in the app and not in the worker script.
 * Controls the flow of messages to and from a `WorkerClient` and a `WorkerController`
 */
var ClientWebWorker = /** @class */ (function () {
    /**
     * Creates a new instance of a `ClientWebWorker`
     * @param workerType the worker class
     * @param isTestClient whether the instance is used for testing which will then mock serialization
     */
    function ClientWebWorker(workerType, isTestClient) {
        var _this = this;
        this.isTestClient = isTestClient;
        this.messageBus = {
            onmessage: function () { },
            postMessage: function (resp) {
                _this.onmessage(new MessageEvent('ClientWebWorker', { data: _this.isTestClient ? _this.serialize(resp) : resp }));
            }
        };
        this.controller = new WorkerController(workerType, this.messageBus);
    }
    Object.defineProperty(ClientWebWorker.prototype, "workerInstance", {
        /**
         * Returns instance of worker class
         */
        get: function () {
            return this.controller.workerInstance;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Message listener for a `WorkerClient`
     */
    ClientWebWorker.prototype.onmessage = function (ev) {
    };
    /**
     * Sends messages triggered from a `WorkerClient` to a `WorkerController`
     */
    ClientWebWorker.prototype.postMessage = function (resp) {
        this.messageBus.onmessage(new MessageEvent('ClientWebWorker', { data: this.isTestClient ? this.serialize(resp) : resp }));
    };
    /**
     * Unsubscribes from all subscriptions in the `WorkerController` and then destroys the controller
     */
    ClientWebWorker.prototype.terminate = function () {
        this.controller.removeAllSubscriptions();
        this.controller = null;
    };
    /**
     * Used for testing to mock the serialization that occurs when native the postMessage or onmessage are used to communicate with a worker script
     * @param obj object to be serialised
     */
    ClientWebWorker.prototype.serialize = function (obj) {
        return JSON.parse(JSON.stringify(obj));
    };
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    ClientWebWorker.prototype.onerror = function (err) {
    };
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    ClientWebWorker.prototype.addEventListener = function () {
    };
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    ClientWebWorker.prototype.removeEventListener = function () {
    };
    /**
     * Ensures class conforms to the native `Worker` class
     * @NotImplemented
     */
    ClientWebWorker.prototype.dispatchEvent = function (evt) {
        return true;
    };
    return ClientWebWorker;
}());
export { ClientWebWorker };
