// angular-web-worker/esm2015/lib/callable-decorator.js
import 'reflect-metadata';
import { WorkerEvents, WorkerUtils, WorkerAnnotations } from 'angular-web-worker/common';
/**
 * Allows the decorated worker method to be called, and its value returned, from the `WorkerClient.call()` method.
 * Can be used on both asynchronous and synchronous methods.
 * @Serialized Functions will not be copied and circular referencing structures will cause errors. This applies to both the function arguments and the value returned by the function
 * @param options Configurable options defining how the decorated method is called from a `WorkerClient`
 */
export function Callable(options) {
    return function (target, propertyKey, descriptor) {
        const opts = { shallowTransfer: false };
        if (options) {
            opts.shallowTransfer = options.shallowTransfer ? true : false;
        }
        WorkerUtils.pushAnnotation(target.constructor, WorkerAnnotations.Callables, {
            name: propertyKey,
            shallowTransfer: opts.shallowTransfer,
            returnType: Reflect.getMetadata('design:returntype', target, propertyKey)
        });
        const originalMethod = descriptor.value;
        descriptor.value = function () {
            const context = this;
            const args = Array.prototype.slice.call(arguments);
            const config = context.__worker_config__;
            if (config) {
                if (config.isClient) {
                    const secret = {
                        clientSecret: context.__worker_config__.clientSecret,
                        type: WorkerEvents.Callable,
                        propertyName: propertyKey,
                        body: {
                            args: args
                        }
                    };
                    return secret;
                }
                else {
                    return originalMethod.call(context, ...args);
                }
            }
            else {
                return originalMethod.call(context, ...args);
            }
        };
        return descriptor;
    };
}
